File: internal/agent/generic.go
package agent

import (
	"context"
	"strings"

	"github.com/a-barwick/agent-crucible/internal/clock"
	"github.com/a-barwick/agent-crucible/internal/schema"
	"github.com/a-barwick/agent-crucible/internal/trace"
)

// Generic runs a pasted Spec: tool schemas + graph (+ optional node bindings).
// Standard closer nodes reuse the CRM patient so the same bugs still fire.
// Custom tool names walk the generic world instead of the CRM switch.
type Generic struct {
	spec  Spec
	clock *clock.Clock
	crm   *CRM
	Model Model
	Saver Checkpointer
}

func NewGeneric(spec Spec, clk *clock.Clock) *Generic {
	if clk == nil {
		clk = clock.New()
	}
	crm := NewCRM(clk)
	return &Generic{spec: spec, clock: clk, crm: crm, Model: crm.Model, Saver: crm.Saver}
}

// NewFromSpec builds an in-process agent from a pasted contract.
func NewFromSpec(spec Spec, clk *clock.Clock) Agent {
	if spec.Name == "" {
		spec.Name = "pasted"
	}
	if spec.Framework == "" {
		spec.Framework = "generic"
	}
	if spec.Graph.Start == "" && len(spec.Graph.Nodes) == 0 {
		if len(spec.Tools) > 0 && !looksLikeCRM(spec.Tools) {
			spec.Graph = graphFromTools(spec.Tools)
		} else {
			spec.Graph = CRMGraphSpec()
			if len(spec.Tools) == 0 {
				spec.Tools = append([]schema.Tool(nil), CRMTools()...)
			}
		}
	}
	return NewGeneric(spec, clk)
}

func (a *Generic) Spec() Spec { return a.spec }

func (a *Generic) Run(ctx context.Context, st State, bus Bus, rec *trace.Recorder, hook Hook) (Result, error) {
	if a.Saver != nil {
		a.crm.Saver = a.Saver
	}
	if a.Model != nil {
		a.crm.Model = a.Model
	}
	if st.Objective == "" && a.spec.Objective != "" {
		st.Objective = a.spec.Objective
	}
	if len(st.Companies) == 0 && len(a.spec.Companies) > 0 {
		st.Companies = a.spec.Companies
	}
	if st.ThreadID == "" {
		st.ThreadID = "pasted"
	}
	if st.Intent.EntityName() == "" && st.Objective != "" {
		st.Intent = ParseIntentWith(st.Objective, st.Companies)
	}
	nodes := map[string]NodeFunc{}
	for _, name := range a.spec.Graph.Nodes {
		if name == "end" || name == "abort" {
			continue
		}
		bind := a.spec.NodeTools[name]
		if bind.Kind == "" {
			bind = inferBind(name, a.spec)
		}
		nodes[name] = a.remap(name, a.makeNode(name, bind))
	}
	start := a.spec.Graph.Start
	if start == "" {
		start = "plan"
	}
	g := &Graph{
		Name:         a.spec.Name,
		Start:        start,
		Nodes:        nodes,
		MaxSteps:     20,
		Clock:        a.clock,
		Checkpointer: a.crm.Saver,
	}
	return g.Run(ctx, &st, bus, rec, hook)
}

func (a *Generic) makeNode(name string, bind NodeBinding) NodeFunc {
	switch bind.Kind {
	case "plan":
		return a.crm.plan
	case "lookup":
		return a.crm.lookup
	case "fetch":
		return a.crm.fetch
	case "enrich":
		return a.crm.enrich
	case "authorize":
		return a.crm.authorize
	case "write":
		return a.crm.write
	case "notify":
		return a.crm.notify
	case "tool":
		return a.toolNode(name, bind)
	default:
		return a.crm.plan
	}
}

func (a *Generic) remap(name string, fn NodeFunc) NodeFunc {
	return func(ctx context.Context, st *State, bus Bus, rec *trace.Recorder) (string, error) {
		next, err := fn(ctx, st, bus, rec)
		if err != nil {
			return next, err
		}
		if next == "" || next == "end" || next == "abort" {
			return next, nil
		}
		if a.hasNode(next) {
			return next, nil
		}
		return nextFrom(a.spec.Graph, name, rec), nil
	}
}

func (a *Generic) hasNode(name string) bool {
	for _, n := range a.spec.Graph.Nodes {
		if n == name {
			return true
		}
	}
	return false
}

func (a *Generic) toolNode(name string, bind NodeBinding) NodeFunc {
	return func(ctx context.Context, st *State, bus Bus, rec *trace.Recorder) (string, error) {
		args := inferArgs(st, bind.Tool, a.spec)
		for arg, path := range bind.ArgsFrom {
			args[arg] = stateValue(st, path)
		}
		hijackReadArgs(st, bind.Tool, args, rec)
		res, err := callRetry(ctx, bus, rec, bind.Tool, args)
		if err != nil {
			return "abort", err
		}
		if res.IsTransportError() {
			st.LastError = res.Error
			return "abort", nil
		}
		applySaves(st, res.Data, bind.Save)
		overwriteFromMemory(st, bind.Tool, rec)
		if schema.IsWriteLike(bind.Tool) {
			// Same bug as the CRM write node: a non-timeout envelope is "done".
			st.Wrote = true
			if s := schema.StringField(res.Data, "status"); s != "" {
				st.Status = s
			} else if s := schema.StringField(args, "status"); s != "" {
				st.Status = s
			}
			if res.Error == "permission_denied" {
				rec.State("write ignored permission_denied", map[string]any{"tool": bind.Tool})
			}
			if res.OK && (res.Data == nil || len(res.Data) == 0) {
				rec.State("write accepted empty success payload", map[string]any{"tool": bind.Tool})
			}
		}
		if schema.IsEmailLike(bind.Tool) {
			st.Notified = res.OK || res.Error == ""
		}
		return nextFrom(a.spec.Graph, name, rec), nil
	}
}

func hijackReadArgs(st *State, tool string, args map[string]any, rec *trace.Recorder) {
	if schema.IsWriteLike(tool) || schema.IsEmailLike(tool) || schema.IsPermissionLike(tool) {
		return
	}
	if st.Junk == "" {
		return
	}
	hijack := lastCompany(st.Junk, st.Companies)
	if hijack == "" || hijack == st.Intent.EntityName() {
		return
	}
	changed := false
	for _, k := range []string{"query", "company", "name", "title"} {
		if _, ok := args[k]; ok {
			args[k] = hijack
			changed = true
		}
	}
	if changed {
		rec.State("lookup hijacked by context ballast", map[string]any{"company": hijack, "tool": tool})
	}
}

// overwriteFromMemory is the stale-memory bug in the generic runner: a populated
// checkpoint beats whatever the read just returned. It only logs when it changed
// something, so a graph with four read nodes reports one stale-memory event
// rather than four identical ones.
func overwriteFromMemory(st *State, tool string, rec *trace.Recorder) {
	if schema.IsWriteLike(tool) || schema.IsEmailLike(tool) || schema.IsPermissionLike(tool) {
		return
	}
	id := st.Memory.TargetID()
	if id == "" {
		return
	}
	changed := st.DealID != id || st.RecordID != id
	st.DealID = id
	st.RecordID = id
	if s := st.Memory.DealStatus; s != "" && st.Status != s {
		st.Status = s
		changed = true
	}
	if a := st.Memory.Amount; a != 0 && st.Amount != a {
		st.Amount = a
		changed = true
	}
	if !changed {
		return
	}
	rec.State("enrich trusted stale memory", map[string]any{"deal_id": st.DealID, "record_id": id, "tool": tool})
}

func inferBind(name string, spec Spec) NodeBinding {
	switch name {
	case "plan":
		return NodeBinding{Kind: "plan"}
	case "lookup":
		return NodeBinding{Kind: "lookup", Tool: "lookup_contact"}
	case "fetch":
		return NodeBinding{Kind: "fetch", Tool: "get_deal"}
	case "enrich":
		return NodeBinding{Kind: "enrich"}
	case "authorize":
		return NodeBinding{Kind: "authorize", Tool: "check_permission"}
	case "write":
		return NodeBinding{Kind: "write", Tool: "write_deal"}
	case "notify":
		return NodeBinding{Kind: "notify", Tool: "send_email"}
	}
	for _, t := range spec.Tools {
		if t.Name == name || strings.Contains(name, t.Name) || strings.Contains(t.Name, name) {
			return NodeBinding{Kind: "tool", Tool: t.Name}
		}
	}
	want := schema.Classify(name)
	if want != schema.KindRead || schema.IsWriteLike(name) || schema.IsEmailLike(name) || schema.IsPermissionLike(name) {
		for _, t := range spec.Tools {
			if schema.Classify(t.Name) == want {
				return NodeBinding{Kind: "tool", Tool: t.Name}
			}
		}
	}
	return NodeBinding{Kind: "plan"}
}

func inferArgs(st *State, tool string, spec Spec) map[string]any {
	args := map[string]any{}
	t, ok := schema.Find(spec.Tools, tool)
	required := t.Required
	if !ok || len(required) == 0 {
		required = defaultArgNames(tool)
	}
	for _, name := range required {
		if v, ok := argFromState(st, name, schema.Classify(tool)); ok {
			args[name] = v
		}
	}
	if schema.IsWriteLike(tool) {
		if s := ActionStatus(st.Intent.ActionName()); s != "" {
			if _, exists := args["status"]; !exists || schema.StringField(args, "status") == "" {
				args["status"] = s
			}
		}
	}
	return args
}

func defaultArgNames(tool string) []string {
	switch schema.Classify(tool) {
	case schema.KindWrite:
		return []string{"id", "status"}
	case schema.KindEmail:
		return []string{"to", "subject", "body"}
	case schema.KindPermission:
		return []string{"perm"}
	default:
		return []string{"query", "company", "id"}
	}
}

func argFromState(st *State, name string, kind schema.Kind) (any, bool) {
	switch name {
	case "company", "query", "name", "title", "entity":
		if st.Intent.EntityName() != "" {
			return st.Intent.EntityName(), true
		}
	case "id", "record_id", "ticket_id", "deal_id":
		if id := st.TargetID(); id != "" {
			return id, true
		}
	case "contact_id":
		if st.ContactID != "" {
			return st.ContactID, true
		}
	case "status":
		if kind == schema.KindWrite {
			if s := ActionStatus(st.Intent.ActionName()); s != "" {
				return s, true
			}
		}
		if st.Status != "" {
			return st.Status, true
		}
	case "to", "email", "recipient":
		if st.AE != "" {
			return st.AE, true
		}
	case "perm", "permission":
		return "crm.write", true
	case "amount":
		return st.Amount, true
	case "owner_id":
		if st.OwnerID != "" {
			return st.OwnerID, true
		}
	case "close_date":
		if st.CloseDate != "" {
			return st.CloseDate, true
		}
	case "subject":
		return "update: " + st.Intent.EntityName(), true
	case "body", "text":
		return "deal=" + st.DealID + " status=" + st.Status, true
	}
	return nil, false
}

func applySaves(st *State, data map[string]any, save map[string]string) {
	if len(save) == 0 {
		save = map[string]string{
			"id": "deal_id", "status": "status", "ae": "ae",
			"email": "ae", "amount": "amount", "owner_id": "owner_id",
			"contact_id": "contact_id",
		}
	}
	for from, to := range save {
		if data == nil {
			continue
		}
		if _, ok := data[from]; !ok {
			continue
		}
		setState(st, to, schema.StringField(data, from))
		if to == "amount" {
			st.Amount = schema.IntField(data, from)
		}
	}
}

func stateValue(st *State, path string) any {
	switch path {
	case "intent.company", "intent.entity", "company", "entity", "query", "name":
		return st.Intent.EntityName()
	case "contact_id":
		return st.ContactID
	case "deal_id", "id", "record_id", "ticket_id":
		return st.TargetID()
	case "ae":
		return st.AE
	case "status":
		return st.Status
	case "amount":
		return st.Amount
	case "owner_id":
		return st.OwnerID
	case "close_date":
		return st.CloseDate
	default:
		return path
	}
}

func setState(st *State, field, val string) {
	switch field {
	case "contact_id":
		st.ContactID = val
	case "ae":
		st.AE = val
	case "deal_id", "id", "record_id", "ticket_id":
		st.DealID = val
		st.RecordID = val
	case "status":
		st.Status = val
	case "owner_id":
		st.OwnerID = val
	case "close_date":
		st.CloseDate = val
	}
}

// nextFrom picks the outgoing edge to follow. A pasted graph may declare
// several edges out of one node; without a condition to evaluate, taking the
// first non-abort edge silently drops the rest, so say so in the timeline
// rather than leaving a node the user declared permanently unreachable.
func nextFrom(g GraphSpec, from string, rec *trace.Recorder) string {
	var outs []string
	for _, e := range g.Edges {
		if e.From == from && e.To != "abort" {
			outs = append(outs, e.To)
		}
	}
	if len(outs) == 0 {
		return "end"
	}
	if len(outs) > 1 && rec != nil {
		rec.State("graph has an unconditional branch; took the first edge", map[string]any{
			"node": from, "took": outs[0], "skipped": outs[1:],
		})
	}
	return outs[0]
}

func looksLikeCRM(tools []schema.Tool) bool {
	return schema.LooksLikeCRM(tools)
}

// LooksLikeCRM reports whether the spec is the sample closer's tool set.
func LooksLikeCRM(tools []schema.Tool) bool {
	return schema.LooksLikeCRM(tools)
}

func graphFromTools(tools []schema.Tool) GraphSpec {
	nodes := []string{"plan"}
	edges := []Edge{}
	prev := "plan"
	for _, t := range tools {
		nodes = append(nodes, t.Name)
		edges = append(edges, Edge{From: prev, To: t.Name})
		prev = t.Name
	}
	edges = append(edges, Edge{From: prev, To: "end"})
	nodes = append(nodes, "end", "abort")
	return GraphSpec{Start: "plan", Nodes: nodes, Edges: edges}
}

File: runtime/crucible_rt/generic.py
"""Compile a pasted spec into a LangGraph StateGraph or an ADK-shaped walk.

The chamber stays on the other side of the tools. This module only
decides node order and which tool name to callback.
"""

from typing import Any

from . import patient
from .intent import last_company, parse_model_intent

CRM_TOOLS = {"lookup_contact", "get_deal", "write_deal", "send_email", "check_permission"}
CRM_NODES = {"plan", "lookup", "fetch", "enrich", "authorize", "write", "notify", "end", "abort"}


def should_compile(spec: dict | None) -> bool:
    """True when the sidecar should build from the pasted spec, not the CRM closer."""
    if not spec:
        return False
    tools = spec.get("tools") or []
    graph = spec.get("graph") or {}
    nodes = list(graph.get("nodes") or [])
    if looks_like_crm(tools) and (not nodes or set(nodes) <= CRM_NODES):
        return False
    return bool(tools) or bool(nodes)


def looks_like_crm(tools) -> bool:
    for t in tools or []:
        name = t.get("name") if isinstance(t, dict) else t
        if name in CRM_TOOLS:
            return True
    return False


def classify(name: str) -> str:
    n = (name or "").lower()
    if any(p in n for p in ("email", "notify", "mail", "send_message")):
        return "email"
    if any(p in n for p in ("permission", "authorize", "acl")):
        return "permission"
    if any(p in n for p in ("write", "update", "patch", "create", "delete", "refund", "upsert", "set_")):
        return "write"
    return "read"


def is_write(name: str) -> bool:
    return classify(name) == "write"


def is_email(name: str) -> bool:
    return classify(name) == "email"


def is_permission(name: str) -> bool:
    return classify(name) == "permission"


def action_status(action: str) -> str:
    return {
        "close_won": "Closed-Won",
        "on_hold": "On-Hold",
        "refund": "Refunded",
        "resolve": "Resolved",
    }.get(action or "", "")


def graph_from_spec(spec: dict) -> dict:
    graph = spec.get("graph") or {}
    if graph.get("start") or graph.get("nodes"):
        return {
            "start": graph.get("start") or "plan",
            "nodes": list(graph.get("nodes") or []),
            "edges": list(graph.get("edges") or []),
        }
    tools = spec.get("tools") or []
    nodes = ["plan"]
    edges = []
    prev = "plan"
    for t in tools:
        name = t.get("name") if isinstance(t, dict) else str(t)
        nodes.append(name)
        edges.append({"from": prev, "to": name})
        prev = name
    edges.append({"from": prev, "to": "end"})
    nodes += ["end", "abort"]
    return {"start": "plan", "nodes": nodes, "edges": edges}


def next_from(graph: dict, frm: str) -> str:
    for e in graph.get("edges") or []:
        if e.get("from") == frm and e.get("to") != "abort":
            return e.get("to") or "end"
    return "end"


def infer_bind(name: str, spec: dict) -> dict:
    binds = spec.get("node_tools") or {}
    if name in binds and isinstance(binds[name], dict):
        b = dict(binds[name])
        if b.get("kind") or b.get("tool"):
            return b
    if name == "plan":
        return {"kind": "plan"}
    if name == "lookup":
        return {"kind": "lookup", "tool": "lookup_contact"}
    if name == "fetch":
        return {"kind": "fetch", "tool": "get_deal"}
    if name == "enrich":
        return {"kind": "enrich"}
    if name == "authorize":
        return {"kind": "authorize", "tool": "check_permission"}
    if name == "write":
        return {"kind": "write", "tool": "write_deal"}
    if name == "notify":
        return {"kind": "notify", "tool": "send_email"}
    for t in spec.get("tools") or []:
        tn = t.get("name") if isinstance(t, dict) else str(t)
        if tn == name or name in tn or tn in name:
            return {"kind": "tool", "tool": tn}
    want = classify(name)
    if want != "read" or is_write(name) or is_email(name) or is_permission(name):
        for t in spec.get("tools") or []:
            tn = t.get("name") if isinstance(t, dict) else str(t)
            if classify(tn) == want:
                return {"kind": "tool", "tool": tn}
    return {"kind": "plan"}


def find_tool(spec: dict, name: str) -> dict:
    for t in spec.get("tools") or []:
        if isinstance(t, dict) and t.get("name") == name:
            return t
    return {}


def infer_args(state: dict, tool: str, spec: dict) -> dict:
    args: dict[str, Any] = {}
    t = find_tool(spec, tool)
    required = list(t.get("required") or [])
    if not required:
        required = default_arg_names(tool)
    kind = classify(tool)
    for name in required:
        v = arg_from_state(state, name, kind)
        if v is not None and v != "":
            args[name] = v
    if is_write(tool):
        intent = state.get("intent") or {}
        status = action_status(intent.get("action") or intent.get("deal_action") or "")
        if status and not args.get("status"):
            args["status"] = status
    return args


def default_arg_names(tool: str) -> list[str]:
    kind = classify(tool)
    if kind == "write":
        return ["id", "status"]
    if kind == "email":
        return ["to", "subject", "body"]
    if kind == "permission":
        return ["perm"]
    return ["query", "company", "id"]


def arg_from_state(state: dict, name: str, kind: str):
    intent = state.get("intent") or {}
    if name in ("company", "query", "name", "title"):
        return intent.get("entity") or intent.get("company") or None
    if name in ("id", "record_id", "ticket_id", "deal_id"):
        return state.get("record_id") or state.get("deal_id") or None
    if name == "contact_id":
        return state.get("contact_id") or None
    if name == "status":
        if kind == "write":
            s = action_status((intent.get("action") or intent.get("deal_action") or ""))
            if s:
                return s
        return state.get("status") or None
    if name in ("to", "email", "recipient"):
        return state.get("ae") or None
    if name in ("perm", "permission"):
        return "crm.write"
    if name == "amount":
        return state.get("amount") or 0
    if name == "owner_id":
        return state.get("owner_id") or None
    if name == "close_date":
        return state.get("close_date") or None
    if name == "subject":
        return "update: " + (intent.get("company") or "")
    if name in ("body", "text"):
        return "deal=" + str(state.get("deal_id") or "") + " status=" + str(state.get("status") or "")
    return None


def apply_saves(state: dict, data: dict | None, save: dict | None) -> None:
    save = save or {
        "id": "deal_id",
        "status": "status",
        "ae": "ae",
        "email": "ae",
        "amount": "amount",
        "owner_id": "owner_id",
        "contact_id": "contact_id",
    }
    if not data:
        return
    for src, dest in save.items():
        if src not in data:
            continue
        val = data.get(src)
        if dest == "amount":
            try:
                state["amount"] = int(val or 0)
            except (TypeError, ValueError):
                state["amount"] = 0
            continue
        state[dest if dest != "id" else "deal_id"] = "" if val is None else str(val)
        if dest in ("deal_id", "id", "record_id", "ticket_id"):
            state["deal_id"] = "" if val is None else str(val)


def state_value(state: dict, path: str):
    intent = state.get("intent") or {}
    if path in ("intent.company", "intent.entity", "company", "entity", "query", "name"):
        return intent.get("entity") or intent.get("company")
    if path == "contact_id":
        return state.get("contact_id")
    if path in ("deal_id", "id", "record_id", "ticket_id"):
        return state.get("record_id") or state.get("deal_id")
    if path == "ae":
        return state.get("ae")
    if path == "status":
        return state.get("status")
    if path == "amount":
        return state.get("amount")
    if path == "owner_id":
        return state.get("owner_id")
    if path == "close_date":
        return state.get("close_date")
    return path


def make_node_fn(cb, name: str, spec: dict):
    bind = infer_bind(name, spec)
    kind = bind.get("kind") or "tool"

    def fn(state: dict) -> dict:
        if kind == "plan":
            return _plan(cb, state, spec)
        if kind == "enrich":
            state = patient.merge_hook(state, cb.before(name))
            state["steps"] = int(state.get("steps") or 0) + 1
            return state
        return _tool(cb, state, name, bind, spec)

    return fn


def _plan(cb, state: dict, spec: dict) -> dict:
    from .model import CloserPlanner

    state = patient.merge_hook(state, cb.before("plan"))
    companies = list(state.get("companies") or spec.get("companies") or [])
    model = CloserPlanner(companies=companies, partial=bool(state.get("partial")))
    msg = model.invoke(state.get("objective") or "")
    state["intent"] = parse_model_intent(msg.content, state.get("objective") or "", companies)
    state["steps"] = int(state.get("steps") or 0) + 1
    if not state["intent"].get("company"):
        state["last_error"] = "empty_company"
        state["terminal"] = "abort"
    return state


def _tool(cb, state: dict, name: str, bind: dict, spec: dict) -> dict:
    state = patient.merge_hook(state, cb.before(name))
    tool = bind.get("tool") or name
    args = infer_args(state, tool, spec)
    for arg, path in (bind.get("args_from") or {}).items():
        args[arg] = state_value(state, path)
    if classify(tool) == "read" and state.get("junk"):
        hijack = last_company(state.get("junk") or "", state.get("companies"))
        intent = state.get("intent") or {}
        if hijack and hijack != (intent.get("company") or ""):
            for k in ("query", "company", "name", "title"):
                if k in args:
                    args[k] = hijack
            if hasattr(cb, "state"):
                cb.state("lookup hijacked by context ballast", {"company": hijack, "tool": tool})
    res = cb.retry_tool(tool, args)
    state["steps"] = int(state.get("steps") or 0) + 1
    if patient.transport(res):
        state["last_error"] = res.get("error") or "timeout"
        state["terminal"] = "abort"
        return state
    d = patient.data(res)
    apply_saves(state, d, bind.get("save"))
    mem = state.get("memory") or {}
    mid = (mem.get("record_id") or mem.get("deal_id") or "") if classify(tool) == "read" else ""
    if mid:
        state["deal_id"] = mid
        state["record_id"] = mid
        if mem.get("deal_status"):
            state["status"] = mem["deal_status"]
        if hasattr(cb, "state"):
            cb.state("enrich trusted stale memory", {"deal_id": mid, "record_id": mid, "tool": tool})
    if is_write(tool):
        # Same bug as the CRM write node: a non-timeout envelope is "done".
        state["wrote"] = True
        status = d.get("status") or args.get("status") or ""
        if status:
            state["status"] = status
        if res.get("error") == "permission_denied" and hasattr(cb, "state"):
            cb.state("write ignored permission_denied", {"tool": tool})
        if res.get("ok") and not d and hasattr(cb, "state"):
            cb.state("write accepted empty success payload", {"tool": tool})
    if is_email(tool):
        state["notified"] = bool(res.get("ok") or not res.get("error"))
    return state


def walk(cb, state: dict, spec: dict) -> dict:
    gspec = graph_from_spec(spec)
    node = gspec.get("start") or "plan"
    steps = 0
    while node and node not in ("end", "abort"):
        if state.get("terminal") == "abort":
            break
        fn = make_node_fn(cb, node, spec)
        state = fn(state)
        if state.get("terminal") == "abort":
            break
        node = next_from(gspec, node)
        steps += 1
        if steps >= 20:
            state["last_error"] = "max_steps"
            state["terminal"] = "abort"
            break
    if not state.get("terminal"):
        state["terminal"] = "end"
    return state


def run_adk(cb, req: dict) -> dict:
    spec = req.get("spec") or {}
    state = patient.seed(req)
    if spec.get("companies") and not state.get("companies"):
        state["companies"] = list(spec["companies"])
    state = walk(cb, state, spec)
    out = patient.finish(state, "adk")
    tools = []
    for t in spec.get("tools") or []:
        tools.append(t.get("name") if isinstance(t, dict) else str(t))
    out["adk"] = {
        "agent": spec.get("name") or "pasted",
        "instruction": spec.get("description") or "Follow the objective. Use the pasted tools in order.",
        "tools": tools,
        "generic": True,
        "session": req.get("thread_id") or "adk",
    }
    return out


def run_langgraph(cb, req: dict) -> dict:
    from typing import TypedDict

    from langgraph.graph import END, START, StateGraph

    try:
        from langgraph.checkpoint.memory import InMemorySaver
    except ImportError:  # pragma: no cover
        from langgraph.checkpoint.memory import MemorySaver as InMemorySaver

    class AgentState(TypedDict, total=False):
        objective: str
        intent: dict
        memory: dict
        junk: str
        companies: list
        partial: bool
        contact_id: str
        ae: str
        deal_id: str
        status: str
        amount: int
        close_date: str
        owner_id: str
        permitted: bool
        wrote: bool
        notified: bool
        last_error: str
        terminal: str
        steps: int

    spec = req.get("spec") or {}
    gspec = graph_from_spec(spec)

    def wrap(fn):
        def node(state: AgentState) -> dict:
            nxt = fn(dict(state))
            return {k: nxt[k] for k in nxt}

        return node

    def route_to(default: str):
        def _r(state: AgentState) -> str:
            if state.get("terminal") == "abort":
                return END
            return default

        return _r

    g = StateGraph(AgentState)
    usable = [n for n in gspec["nodes"] if n not in ("end", "abort")]
    if not usable:
        usable = ["plan"]
    for name in usable:
        g.add_node(name, wrap(make_node_fn(cb, name, spec)))
    start = gspec.get("start") or usable[0]
    g.add_edge(START, start)
    for name in usable:
        nxt = next_from(gspec, name)
        if nxt in ("", "end", "abort") or nxt not in usable:
            g.add_conditional_edges(name, route_to(END))
        else:
            g.add_conditional_edges(name, route_to(nxt))
    graph = g.compile(checkpointer=InMemorySaver())
    seed = patient.seed(req)
    if spec.get("companies") and not seed.get("companies"):
        seed["companies"] = list(spec["companies"])
    result = graph.invoke(seed, {"configurable": {"thread_id": req.get("thread_id") or "t"}})
    out = patient.finish(result, "langgraph")
    out["checkpoint"] = True
    out["generic"] = True
    return out
